lru_ttl.go

```diff
@@ -0,0 +1,170 @@
+	}
+
+	lru, err := simplelru.NewLRU(size,
+		func(k interface{}, v interface{}) {
+			if onEvict != nil {
+				onEvict(k, v.(cacheValue).value)
+			}
+		})
+	if err != nil {
+		return nil, err
+	}
+
+	lruWithTTL := &CacheWithTTL{LRU: lru, TTL: ttl}
+
+	// clean expired items
+	go lruWithTTL.cleanup()
+
+	return lruWithTTL, nil
+}
+
+// Add adds the item to the cache. It also includes the `lastAccessTime` to the value.
+// Life of an item can be increased by calling `Add` multiple times on the same key.
+func (c *CacheWithTTL) Add(key, value interface{}) bool {
+	c.lock.Lock()
+	defer c.lock.Unlock()
+	return c.LRU.Add(key,
+		cacheValue{
+			value:          value,
+			lastAccessTime: time.Now(),
+		})
+}
+
+// Get looks up a key's value from the cache.
+// Also, it unmarshals `lastAccessTime` from `Get` response
+func (c *CacheWithTTL) Get(key interface{}) (value interface{}, ok bool) {
+	c.lock.Lock()
+	defer c.lock.Unlock()
+
+	val, ok := c.LRU.Get(key)
+	// while the cleanup routine is catching will the other items, remove the item
+	// if someone tries to access it through this GET call.
+	if ok {
+		if time.Now().After(val.(cacheValue).lastAccessTime.Add(c.TTL)) {
+			c.LRU.Remove(key)
+		} else {
+			return val.(cacheValue).value, ok
+		}
+	}
+
+	return nil, false
+}
+
+// Peek returns the key value (or undefined if not found) without updating
+// the "recently used"-ness of the key.
+// Also, it unmarshals the `lastAccessTime` from the result
+func (c *CacheWithTTL) Peek(key interface{}) (value interface{}, ok bool) {
+	c.lock.Lock()
+	defer c.lock.Unlock()
+
+	val, ok := c.LRU.Peek(key)
+	if ok {
+		return val.(cacheValue).value, ok
+	}
+	return val, ok
+}
+
+// Contains checks if a key is in the cache, without updating the
+// recent-ness or deleting it for being stale.
+func (c *CacheWithTTL) Contains(key interface{}) bool {
+	c.lock.RLock()
+	defer c.lock.RUnlock()
+	return c.LRU.Contains(key)
+}
+
+// Purge is used to completely clear the cache.
+func (c *CacheWithTTL) Purge() {
+	c.lock.Lock()
+	c.LRU.Purge()
+	c.lock.Unlock()
+}
+
+// Remove removes the provided key from the cache.
+func (c *CacheWithTTL) Remove(key interface{}) bool {
+	c.lock.Lock()
+	defer c.lock.Unlock()
+	return c.LRU.Remove(key)
+}
+
+// RemoveOldest removes the oldest item from the cache.
+func (c *CacheWithTTL) RemoveOldest() (key interface{}, value interface{}, ok bool) {
+	c.lock.Lock()
+	defer c.lock.Unlock()
+	return c.LRU.RemoveOldest()
+}
+
+// Keys returns a slice of the keys in the cache, from oldest to newest.
+func (c *CacheWithTTL) Keys() []interface{} {
+	c.lock.RLock()
+	defer c.lock.RUnlock()
+	return c.LRU.Keys()
+}
+
+// Len returns the number of items in the cache.
+func (c *CacheWithTTL) Len() int {
+	c.lock.RLock()
+	defer c.lock.RUnlock()
+	return c.LRU.Len()
+}
+
+// cleanup deletes all the expired items
+func (c *CacheWithTTL) cleanup() {
+	ticker := time.NewTicker(2 * time.Millisecond)
```

Comment:
A heap would be a more efficient way to do this cleanup.  Also every 2 ms is very frequently.